File: sdk/lx-trading-py/python/lx_trading/math.py
"""
Financial mathematics for accurate market pricing.

Includes:
- Options pricing (Black-Scholes, Greeks)
- AMM pricing (Constant Product, Concentrated Liquidity)
- Risk metrics (VaR, CVaR, Sharpe, Sortino)
- Statistical measures (volatility, drawdown)
"""

from decimal import Decimal
from typing import List, Optional, Tuple, Union
import math

# Use numpy for performance if available
try:
    import numpy as np
    from scipy import stats
    from scipy.optimize import brentq
    HAS_NUMPY = True
except ImportError:
    HAS_NUMPY = False


# =============================================================================
# Options Pricing
# =============================================================================

def black_scholes(
    S: float,  # Spot price
    K: float,  # Strike price
    T: float,  # Time to expiry (years)
    r: float,  # Risk-free rate
    sigma: float,  # Volatility
    option_type: str = "call",  # "call" or "put"
) -> float:
    """
    Black-Scholes option pricing.

    Args:
        S: Current spot price
        K: Strike price
        T: Time to expiration in years
        r: Risk-free interest rate (annualized)
        sigma: Volatility (annualized)
        option_type: "call" or "put"

    Returns:
        Option price

    Example:
        >>> price = black_scholes(100, 100, 1, 0.05, 0.2, "call")
        >>> print(f"Call price: ${price:.2f}")
    """
    if T <= 0:
        # At expiry
        if option_type == "call":
            return max(S - K, 0)
        return max(K - S, 0)

    if HAS_NUMPY:
        d1 = (np.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
        d2 = d1 - sigma * np.sqrt(T)

        if option_type == "call":
            return S * stats.norm.cdf(d1) - K * np.exp(-r * T) * stats.norm.cdf(d2)
        return K * np.exp(-r * T) * stats.norm.cdf(-d2) - S * stats.norm.cdf(-d1)

    # Pure Python fallback
    d1 = (math.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * math.sqrt(T))
    d2 = d1 - sigma * math.sqrt(T)

    if option_type == "call":
        return S * _norm_cdf(d1) - K * math.exp(-r * T) * _norm_cdf(d2)
    return K * math.exp(-r * T) * _norm_cdf(-d2) - S * _norm_cdf(-d1)


def implied_volatility(
    price: float,
    S: float,
    K: float,
    T: float,
    r: float,
    option_type: str = "call",
    tol: float = 1e-6,
    max_iter: int = 100,
) -> float:
    """
    Calculate implied volatility from option price using Newton-Raphson.

    Args:
        price: Observed option price
        S: Spot price
        K: Strike price
        T: Time to expiry
        r: Risk-free rate
        option_type: "call" or "put"

    Returns:
        Implied volatility
    """
    if HAS_NUMPY:
        try:
            return brentq(
                lambda sigma: black_scholes(S, K, T, r, sigma, option_type) - price,
                0.001, 5.0
            )
        except ValueError:
            return float('nan')

    # Newton-Raphson fallback
    sigma = 0.2  # Initial guess

    for _ in range(max_iter):
        bs_price = black_scholes(S, K, T, r, sigma, option_type)
        vega = _vega(S, K, T, r, sigma)

        if abs(vega) < 1e-10:
            break

        diff = bs_price - price
        if abs(diff) < tol:
            return sigma

        sigma -= diff / vega
        sigma = max(0.001, min(sigma, 5.0))

    return sigma


def greeks(
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    option_type: str = "call",
) -> dict:
    """
    Calculate option Greeks.

    Returns:
        Dictionary with delta, gamma, theta, vega, rho
    """
    if T <= 0:
        return {"delta": 0, "gamma": 0, "theta": 0, "vega": 0, "rho": 0}

    sqrt_T = math.sqrt(T)
    d1 = (math.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * sqrt_T)
    d2 = d1 - sigma * sqrt_T

    if HAS_NUMPY:
        pdf_d1 = stats.norm.pdf(d1)
        cdf_d1 = stats.norm.cdf(d1)
        cdf_d2 = stats.norm.cdf(d2)
        cdf_neg_d1 = stats.norm.cdf(-d1)
        cdf_neg_d2 = stats.norm.cdf(-d2)
    else:
        pdf_d1 = _norm_pdf(d1)
        cdf_d1 = _norm_cdf(d1)
        cdf_d2 = _norm_cdf(d2)
        cdf_neg_d1 = _norm_cdf(-d1)
        cdf_neg_d2 = _norm_cdf(-d2)

    if option_type == "call":
        delta = cdf_d1
        theta = (
            -S * pdf_d1 * sigma / (2 * sqrt_T)
            - r * K * math.exp(-r * T) * cdf_d2
        )
        rho = K * T * math.exp(-r * T) * cdf_d2
    else:
        delta = cdf_d1 - 1
        theta = (
            -S * pdf_d1 * sigma / (2 * sqrt_T)
            + r * K * math.exp(-r * T) * cdf_neg_d2
        )
        rho = -K * T * math.exp(-r * T) * cdf_neg_d2

    gamma = pdf_d1 / (S * sigma * sqrt_T)
    vega = S * pdf_d1 * sqrt_T / 100  # Per 1% change in vol
    theta = theta / 365  # Daily theta

    return {
        "delta": delta,
        "gamma": gamma,
        "theta": theta,
        "vega": vega,
        "rho": rho,
    }


def _vega(S: float, K: float, T: float, r: float, sigma: float) -> float:
    """Calculate vega for Newton-Raphson."""
    sqrt_T = math.sqrt(T)
    d1 = (math.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * sqrt_T)
    return S * _norm_pdf(d1) * sqrt_T


# =============================================================================
# AMM Pricing
# =============================================================================

def constant_product_price(
    reserve_x: Union[float, Decimal],
    reserve_y: Union[float, Decimal],
    amount_in: Union[float, Decimal],
    fee_rate: float = 0.003,  # 0.3%
    is_x_to_y: bool = True,
) -> Tuple[float, float]:
    """
    Calculate output amount for Uniswap V2 style constant product AMM.

    Formula: x * y = k (invariant)

    Args:
        reserve_x: Reserve of token X
        reserve_y: Reserve of token Y
        amount_in: Amount of input token
        fee_rate: Trading fee (e.g., 0.003 = 0.3%)
        is_x_to_y: True if swapping X for Y

    Returns:
        Tuple of (output_amount, effective_price)

    Example:
        >>> out, price = constant_product_price(1000, 1000, 10, fee_rate=0.003)
        >>> print(f"Output: {out:.4f}, Price: {price:.4f}")
    """
    reserve_x = float(reserve_x)
    reserve_y = float(reserve_y)
    amount_in = float(amount_in)

    amount_in_with_fee = amount_in * (1 - fee_rate)

    if is_x_to_y:
        amount_out = (reserve_y * amount_in_with_fee) / (reserve_x + amount_in_with_fee)
    else:
        amount_out = (reserve_x * amount_in_with_fee) / (reserve_y + amount_in_with_fee)

    effective_price = amount_out / amount_in if amount_in > 0 else 0

    return amount_out, effective_price


def concentrated_liquidity_price(
    liquidity: float,
    sqrt_price_current: float,
    sqrt_price_lower: float,
    sqrt_price_upper: float,
    amount_in: Union[float, Decimal],
    fee_rate: float = 0.003,
    is_token0_in: bool = True,
) -> Tuple[float, float, float]:
    """
    Calculate output for Uniswap V3 style concentrated liquidity.

    Args:
        liquidity: L value (sqrt(x * y))
        sqrt_price_current: Current sqrt(P) = sqrt(y/x)
        sqrt_price_lower: Lower tick sqrt price
        sqrt_price_upper: Upper tick sqrt price
        amount_in: Input amount
        fee_rate: Trading fee
        is_token0_in: True if swapping token0 for token1

    Returns:
        Tuple of (output_amount, new_sqrt_price, price_impact)
    """
    amount_in = float(amount_in)
    amount_in_with_fee = amount_in * (1 - fee_rate)

    if is_token0_in:
        # Swapping X for Y (price goes up)
        # ΔY = L * (sqrt_P_new - sqrt_P_current)
        # ΔX = L * (1/sqrt_P_current - 1/sqrt_P_new)

        delta_inv_sqrt_p = amount_in_with_fee / liquidity
        new_inv_sqrt_p = 1 / sqrt_price_current - delta_inv_sqrt_p

        if new_inv_sqrt_p <= 0:
            new_sqrt_p = sqrt_price_upper
        else:
            new_sqrt_p = 1 / new_inv_sqrt_p

        new_sqrt_p = min(new_sqrt_p, sqrt_price_upper)
        amount_out = liquidity * (new_sqrt_p - sqrt_price_current)
    else:
        # Swapping Y for X (price goes down)
        delta_sqrt_p = amount_in_with_fee / liquidity
        new_sqrt_p = sqrt_price_current - delta_sqrt_p

        new_sqrt_p = max(new_sqrt_p, sqrt_price_lower)
        amount_out = liquidity * (1 / new_sqrt_p - 1 / sqrt_price_current)

    # Price impact
    old_price = sqrt_price_current ** 2
    new_price = new_sqrt_p ** 2
    price_impact = abs(new_price - old_price) / old_price

    return max(amount_out, 0), new_sqrt_p, price_impact


def calculate_liquidity(
    amount_x: float,
    amount_y: float,
    sqrt_price_current: float,
    sqrt_price_lower: float,
    sqrt_price_upper: float,
) -> float:
    """
    Calculate liquidity (L) for a concentrated liquidity position.

    Args:
        amount_x: Amount of token X to provide
        amount_y: Amount of token Y to provide
        sqrt_price_current: Current sqrt price
        sqrt_price_lower: Lower tick sqrt price
        sqrt_price_upper: Upper tick sqrt price

    Returns:
        Liquidity value L
    """
    if sqrt_price_current <= sqrt_price_lower:
        # Only token X
        return amount_x * sqrt_price_lower * sqrt_price_upper / (sqrt_price_upper - sqrt_price_lower)
    elif sqrt_price_current >= sqrt_price_upper:
        # Only token Y
        return amount_y / (sqrt_price_upper - sqrt_price_lower)
    else:
        # Both tokens
        l_x = amount_x * sqrt_price_current * sqrt_price_upper / (sqrt_price_upper - sqrt_price_current)
        l_y = amount_y / (sqrt_price_current - sqrt_price_lower)
        return min(l_x, l_y)


# =============================================================================
# Risk Metrics
# =============================================================================

def volatility(
    returns: List[float],
    annualize: bool = True,
    periods_per_year: int = 252,
) -> float:
    """
    Calculate historical volatility.

    Args:
        returns: List of period returns
        annualize: Whether to annualize the result
        periods_per_year: Trading periods per year (252 for daily)

    Returns:
        Volatility (standard deviation of returns)
    """
    if len(returns) < 2:
        return 0.0

    if HAS_NUMPY:
        std = np.std(returns, ddof=1)
    else:
        mean = sum(returns) / len(returns)
        variance = sum((r - mean) ** 2 for r in returns) / (len(returns) - 1)
        std = math.sqrt(variance)

    if annualize:
        std *= math.sqrt(periods_per_year)

    return std


def sharpe_ratio(
    returns: List[float],
    risk_free_rate: float = 0.0,
    periods_per_year: int = 252,
) -> float:
    """
    Calculate Sharpe ratio.

    Args:
        returns: List of period returns
        risk_free_rate: Annual risk-free rate
        periods_per_year: Trading periods per year

    Returns:
        Sharpe ratio
    """
    if len(returns) < 2:
        return 0.0

    if HAS_NUMPY:
        mean_return = np.mean(returns)
        std_return = np.std(returns, ddof=1)
    else:
        mean_return = sum(returns) / len(returns)
        variance = sum((r - mean_return) ** 2 for r in returns) / (len(returns) - 1)
        std_return = math.sqrt(variance)

    if std_return == 0:
        return 0.0

    # Annualize
    period_rf = risk_free_rate / periods_per_year
    excess_return = mean_return - period_rf

    return (excess_return * periods_per_year) / (std_return * math.sqrt(periods_per_year))


def sortino_ratio(
    returns: List[float],
    risk_free_rate: float = 0.0,
    target_return: float = 0.0,
    periods_per_year: int = 252,
) -> float:
    """
    Calculate Sortino ratio (uses downside deviation).

    Args:
        returns: List of period returns
        risk_free_rate: Annual risk-free rate
        target_return: Minimum acceptable return
        periods_per_year: Trading periods per year

    Returns:
        Sortino ratio
    """
    if len(returns) < 2:
        return 0.0

    # Calculate downside deviation
    downside_returns = [min(r - target_return, 0) ** 2 for r in returns]

    if HAS_NUMPY:
        downside_std = np.sqrt(np.mean(downside_returns))
        mean_return = np.mean(returns)
    else:
        downside_std = math.sqrt(sum(downside_returns) / len(downside_returns))
        mean_return = sum(returns) / len(returns)

    if downside_std == 0:
        return float('inf') if mean_return > risk_free_rate / periods_per_year else 0.0

    period_rf = risk_free_rate / periods_per_year
    excess_return = mean_return - period_rf

    return (excess_return * periods_per_year) / (downside_std * math.sqrt(periods_per_year))


def max_drawdown(prices: List[float]) -> Tuple[float, int, int]:
    """
    Calculate maximum drawdown.

    Args:
        prices: List of prices/equity values

    Returns:
        Tuple of (max_drawdown, peak_index, trough_index)
    """
    if len(prices) < 2:
        return 0.0, 0, 0

    peak = prices[0]
    peak_idx = 0
    max_dd = 0.0
    max_dd_peak = 0
    max_dd_trough = 0

    for i, price in enumerate(prices):
        if price > peak:
            peak = price
            peak_idx = i

        dd = (peak - price) / peak if peak > 0 else 0

        if dd > max_dd:
            max_dd = dd
            max_dd_peak = peak_idx
            max_dd_trough = i

    return max_dd, max_dd_peak, max_dd_trough


def var(
    returns: List[float],
    confidence: float = 0.95,
    method: str = "historical",
) -> float:
    """
    Calculate Value at Risk.

    Args:
        returns: List of returns
        confidence: Confidence level (e.g., 0.95 for 95%)
        method: "historical" or "parametric"

    Returns:
        VaR as a positive number (potential loss)
    """
    if len(returns) < 10:
        return 0.0

    if method == "historical":
        if HAS_NUMPY:
            return -np.percentile(returns, (1 - confidence) * 100)
        sorted_returns = sorted(returns)
        idx = int(len(sorted_returns) * (1 - confidence))
        return -sorted_returns[idx]

    # Parametric (assumes normal distribution)
    if HAS_NUMPY:
        mean = np.mean(returns)
        std = np.std(returns, ddof=1)
        z = stats.norm.ppf(1 - confidence)
    else:
        mean = sum(returns) / len(returns)
        variance = sum((r - mean) ** 2 for r in returns) / (len(returns) - 1)
        std = math.sqrt(variance)
        z = -1.645 if confidence == 0.95 else -2.326  # Approximate

    return -(mean + z * std)


def cvar(
    returns: List[float],
    confidence: float = 0.95,
) -> float:
    """
    Calculate Conditional Value at Risk (Expected Shortfall).

    Args:
        returns: List of returns
        confidence: Confidence level

    Returns:
        CVaR as a positive number (expected loss beyond VaR)
    """
    if len(returns) < 10:
        return 0.0

    var_value = var(returns, confidence, "historical")

    # Average of returns worse than VaR
    tail_returns = [r for r in returns if r <= -var_value]

    if not tail_returns:
        return var_value

    if HAS_NUMPY:
        return -np.mean(tail_returns)
    return -sum(tail_returns) / len(tail_returns)


# =============================================================================
# Helper Functions
# =============================================================================

def _norm_cdf(x: float) -> float:
    """Standard normal CDF (pure Python)."""
    return 0.5 * (1 + math.erf(x / math.sqrt(2)))


def _norm_pdf(x: float) -> float:
    """Standard normal PDF (pure Python)."""
    return math.exp(-0.5 * x * x) / math.sqrt(2 * math.pi)


def price_to_sqrt_price(price: float) -> float:
    """Convert price to sqrt price for concentrated liquidity."""
    return math.sqrt(price)


def sqrt_price_to_price(sqrt_price: float) -> float:
    """Convert sqrt price to regular price."""
    return sqrt_price ** 2


def tick_to_sqrt_price(tick: int, tick_spacing: int = 60) -> float:
    """Convert tick to sqrt price (Uniswap V3 style)."""
    return 1.0001 ** (tick / 2)


def sqrt_price_to_tick(sqrt_price: float, tick_spacing: int = 60) -> int:
    """Convert sqrt price to nearest tick."""
    tick = int(2 * math.log(sqrt_price) / math.log(1.0001))
    return (tick // tick_spacing) * tick_spacing

File: sdk/lx-trading-rust/src/math.rs
//! Financial mathematics.
//!
//! Options pricing (Black-Scholes), Greeks, AMM pricing, and risk metrics.

use std::f64::consts::{PI, SQRT_2};

/// Square root of 2*PI.
pub const SQRT_2PI: f64 = 2.506628274631000502;

/// Greeks for options.
#[derive(Debug, Clone, Copy, Default)]
pub struct Greeks {
    pub delta: f64,
    pub gamma: f64,
    pub theta: f64,
    pub vega: f64,
    pub rho: f64,
}

/// Standard normal CDF (Abramowitz & Stegun approximation).
#[inline]
pub fn norm_cdf(x: f64) -> f64 {
    const A1: f64 = 0.254829592;
    const A2: f64 = -0.284496736;
    const A3: f64 = 1.421413741;
    const A4: f64 = -1.453152027;
    const A5: f64 = 1.061405429;
    const P: f64 = 0.3275911;

    let sign = if x < 0.0 { -1.0 } else { 1.0 };
    let x = x.abs() / SQRT_2;

    let t = 1.0 / (1.0 + P * x);
    let y = 1.0 - (((((A5 * t + A4) * t) + A3) * t + A2) * t + A1) * t * (-x * x).exp();

    0.5 * (1.0 + sign * y)
}

/// Standard normal PDF.
#[inline]
pub fn norm_pdf(x: f64) -> f64 {
    (-0.5 * x * x).exp() / SQRT_2PI
}

/// Black-Scholes option price.
///
/// # Arguments
/// * `s` - Spot price
/// * `k` - Strike price
/// * `t` - Time to expiry (years)
/// * `r` - Risk-free rate
/// * `sigma` - Volatility
/// * `is_call` - True for call, false for put
pub fn black_scholes(s: f64, k: f64, t: f64, r: f64, sigma: f64, is_call: bool) -> f64 {
    if t <= 0.0 {
        return if is_call {
            (s - k).max(0.0)
        } else {
            (k - s).max(0.0)
        };
    }

    let sqrt_t = t.sqrt();
    let d1 = ((s / k).ln() + (r + 0.5 * sigma * sigma) * t) / (sigma * sqrt_t);
    let d2 = d1 - sigma * sqrt_t;

    if is_call {
        s * norm_cdf(d1) - k * (-r * t).exp() * norm_cdf(d2)
    } else {
        k * (-r * t).exp() * norm_cdf(-d2) - s * norm_cdf(-d1)
    }
}

/// Implied volatility from option price (Newton-Raphson).
///
/// # Arguments
/// * `price` - Market price of option
/// * `s` - Spot price
/// * `k` - Strike price
/// * `t` - Time to expiry (years)
/// * `r` - Risk-free rate
/// * `is_call` - True for call, false for put
/// * `tol` - Tolerance (default 1e-6)
/// * `max_iter` - Maximum iterations (default 100)
pub fn implied_volatility(
    price: f64,
    s: f64,
    k: f64,
    t: f64,
    r: f64,
    is_call: bool,
    tol: Option<f64>,
    max_iter: Option<usize>,
) -> f64 {
    let tol = tol.unwrap_or(1e-6);
    let max_iter = max_iter.unwrap_or(100);

    let mut sigma = 0.2;

    for _ in 0..max_iter {
        let bs_price = black_scholes(s, k, t, r, sigma, is_call);

        // Vega
        let sqrt_t = t.sqrt();
        let d1 = ((s / k).ln() + (r + 0.5 * sigma * sigma) * t) / (sigma * sqrt_t);
        let vega = s * norm_pdf(d1) * sqrt_t;

        if vega.abs() < 1e-10 {
            break;
        }

        let diff = bs_price - price;
        if diff.abs() < tol {
            return sigma;
        }

        sigma -= diff / vega;
        sigma = sigma.clamp(0.001, 5.0);
    }

    sigma
}

/// Calculate all Greeks.
///
/// # Arguments
/// * `s` - Spot price
/// * `k` - Strike price
/// * `t` - Time to expiry (years)
/// * `r` - Risk-free rate
/// * `sigma` - Volatility
/// * `is_call` - True for call, false for put
pub fn greeks(s: f64, k: f64, t: f64, r: f64, sigma: f64, is_call: bool) -> Greeks {
    if t <= 0.0 {
        return Greeks::default();
    }

    let sqrt_t = t.sqrt();
    let d1 = ((s / k).ln() + (r + 0.5 * sigma * sigma) * t) / (sigma * sqrt_t);
    let d2 = d1 - sigma * sqrt_t;

    let pdf_d1 = norm_pdf(d1);
    let cdf_d1 = norm_cdf(d1);
    let cdf_d2 = norm_cdf(d2);
    let cdf_neg_d1 = norm_cdf(-d1);
    let cdf_neg_d2 = norm_cdf(-d2);
    let exp_neg_rt = (-r * t).exp();

    let (delta, theta, rho) = if is_call {
        let delta = cdf_d1;
        let theta = (-s * pdf_d1 * sigma / (2.0 * sqrt_t) - r * k * exp_neg_rt * cdf_d2) / 365.0;
        let rho = k * t * exp_neg_rt * cdf_d2;
        (delta, theta, rho)
    } else {
        let delta = cdf_d1 - 1.0;
        let theta = (-s * pdf_d1 * sigma / (2.0 * sqrt_t) + r * k * exp_neg_rt * cdf_neg_d2) / 365.0;
        let rho = -k * t * exp_neg_rt * cdf_neg_d2;
        (delta, theta, rho)
    };

    let gamma = pdf_d1 / (s * sigma * sqrt_t);
    let vega = s * pdf_d1 * sqrt_t / 100.0; // Per 1% vol change

    Greeks {
        delta,
        gamma,
        theta,
        vega,
        rho,
    }
}

/// Constant product AMM price (Uniswap V2 style).
///
/// Returns (output_amount, effective_price).
///
/// # Arguments
/// * `reserve_x` - Reserve of token X
/// * `reserve_y` - Reserve of token Y
/// * `amount_in` - Amount of input token
/// * `fee_rate` - Fee rate (e.g., 0.003 for 0.3%)
/// * `is_x_to_y` - True if swapping X for Y
pub fn constant_product_price(
    reserve_x: f64,
    reserve_y: f64,
    amount_in: f64,
    fee_rate: f64,
    is_x_to_y: bool,
) -> (f64, f64) {
    let amount_in_with_fee = amount_in * (1.0 - fee_rate);

    let amount_out = if is_x_to_y {
        (reserve_y * amount_in_with_fee) / (reserve_x + amount_in_with_fee)
    } else {
        (reserve_x * amount_in_with_fee) / (reserve_y + amount_in_with_fee)
    };

    let effective_price = if amount_in > 0.0 {
        amount_out / amount_in
    } else {
        0.0
    };

    (amount_out, effective_price)
}

/// Concentrated liquidity price (Uniswap V3 style).
///
/// Returns (output_amount, new_sqrt_price, price_impact).
///
/// # Arguments
/// * `liquidity` - Position liquidity
/// * `sqrt_price_current` - Current sqrt price
/// * `sqrt_price_lower` - Lower bound sqrt price
/// * `sqrt_price_upper` - Upper bound sqrt price
/// * `amount_in` - Amount of input token
/// * `fee_rate` - Fee rate
/// * `is_token0_in` - True if token0 is input
pub fn concentrated_liquidity_price(
    liquidity: f64,
    sqrt_price_current: f64,
    sqrt_price_lower: f64,
    sqrt_price_upper: f64,
    amount_in: f64,
    fee_rate: f64,
    is_token0_in: bool,
) -> (f64, f64, f64) {
    let amount_in_with_fee = amount_in * (1.0 - fee_rate);

    let (new_sqrt_p, amount_out) = if is_token0_in {
        // Swapping X for Y (price goes up)
        let delta_inv_sqrt_p = amount_in_with_fee / liquidity;
        let new_inv_sqrt_p = 1.0 / sqrt_price_current - delta_inv_sqrt_p;

        let new_sqrt_p = if new_inv_sqrt_p <= 0.0 {
            sqrt_price_upper
        } else {
            (1.0 / new_inv_sqrt_p).min(sqrt_price_upper)
        };

        let amount_out = liquidity * (new_sqrt_p - sqrt_price_current);
        (new_sqrt_p, amount_out)
    } else {
        // Swapping Y for X (price goes down)
        let delta_sqrt_p = amount_in_with_fee / liquidity;
        let new_sqrt_p = (sqrt_price_current - delta_sqrt_p).max(sqrt_price_lower);

        let amount_out = liquidity * (1.0 / new_sqrt_p - 1.0 / sqrt_price_current);
        (new_sqrt_p, amount_out)
    };

    let old_price = sqrt_price_current * sqrt_price_current;
    let new_price = new_sqrt_p * new_sqrt_p;
    let price_impact = ((new_price - old_price).abs()) / old_price;

    (amount_out.max(0.0), new_sqrt_p, price_impact)
}

/// Calculate liquidity for concentrated position.
///
/// # Arguments
/// * `amount_x` - Amount of token X
/// * `amount_y` - Amount of token Y
/// * `sqrt_price_current` - Current sqrt price
/// * `sqrt_price_lower` - Lower bound sqrt price
/// * `sqrt_price_upper` - Upper bound sqrt price
pub fn calculate_liquidity(
    amount_x: f64,
    amount_y: f64,
    sqrt_price_current: f64,
    sqrt_price_lower: f64,
    sqrt_price_upper: f64,
) -> f64 {
    if sqrt_price_current <= sqrt_price_lower {
        // Only token X
        amount_x * sqrt_price_lower * sqrt_price_upper / (sqrt_price_upper - sqrt_price_lower)
    } else if sqrt_price_current >= sqrt_price_upper {
        // Only token Y
        amount_y / (sqrt_price_upper - sqrt_price_lower)
    } else {
        // Both tokens
        let l_x =
            amount_x * sqrt_price_current * sqrt_price_upper / (sqrt_price_upper - sqrt_price_current);
        let l_y = amount_y / (sqrt_price_current - sqrt_price_lower);
        l_x.min(l_y)
    }
}

/// Convert price to sqrt price.
#[inline]
pub fn price_to_sqrt_price(price: f64) -> f64 {
    price.sqrt()
}

/// Convert sqrt price to price.
#[inline]
pub fn sqrt_price_to_price(sqrt_price: f64) -> f64 {
    sqrt_price * sqrt_price
}

/// Convert tick to sqrt price.
#[inline]
pub fn tick_to_sqrt_price(tick: i32) -> f64 {
    1.0001_f64.powf(tick as f64 / 2.0)
}

/// Convert sqrt price to tick.
#[inline]
pub fn sqrt_price_to_tick(sqrt_price: f64, tick_spacing: i32) -> i32 {
    let tick = (2.0 * sqrt_price.ln() / 1.0001_f64.ln()) as i32;
    (tick / tick_spacing) * tick_spacing
}

#[cfg(test)]
mod tests {
    use super::*;

    fn approx_eq(a: f64, b: f64, margin: f64) -> bool {
        (a - b).abs() < margin
    }

    #[test]
    fn test_black_scholes_atm_call() {
        let price = black_scholes(100.0, 100.0, 1.0, 0.05, 0.2, true);
        assert!(approx_eq(price, 10.45, 0.1));
    }

    #[test]
    fn test_black_scholes_itm_call() {
        let price = black_scholes(110.0, 100.0, 1.0, 0.05, 0.2, true);
        assert!(approx_eq(price, 17.68, 0.1));
    }

    #[test]
    fn test_black_scholes_otm_put() {
        let price = black_scholes(110.0, 100.0, 1.0, 0.05, 0.2, false);
        assert!(approx_eq(price, 2.80, 0.1));
    }

    #[test]
    fn test_black_scholes_zero_time() {
        let call_price = black_scholes(110.0, 100.0, 0.0, 0.05, 0.2, true);
        assert!(approx_eq(call_price, 10.0, 0.001));

        let put_price = black_scholes(90.0, 100.0, 0.0, 0.05, 0.2, false);
        assert!(approx_eq(put_price, 10.0, 0.001));
    }

    #[test]
    fn test_greeks_delta() {
        let g = greeks(100.0, 100.0, 1.0, 0.05, 0.2, true);
        assert!(approx_eq(g.delta, 0.64, 0.02));
    }

    #[test]
    fn test_greeks_gamma() {
        let g = greeks(100.0, 100.0, 1.0, 0.05, 0.2, true);
        assert!(approx_eq(g.gamma, 0.019, 0.002));
    }

    #[test]
    fn test_greeks_vega() {
        let g = greeks(100.0, 100.0, 1.0, 0.05, 0.2, true);
        assert!(approx_eq(g.vega, 0.38, 0.02));
    }

    #[test]
    fn test_greeks_theta() {
        let g = greeks(100.0, 100.0, 1.0, 0.05, 0.2, true);
        assert!(g.theta < 0.0); // Time decay is negative
    }

    #[test]
    fn test_implied_volatility() {
        let true_vol = 0.25;
        let price = black_scholes(100.0, 100.0, 0.5, 0.05, true_vol, true);
        let iv = implied_volatility(price, 100.0, 100.0, 0.5, 0.05, true, None, None);
        assert!(approx_eq(iv, true_vol, 0.01));
    }

    #[test]
    fn test_constant_product_basic() {
        let (out, price) = constant_product_price(1000.0, 1000.0, 10.0, 0.003, true);
        assert!(approx_eq(out, 9.88, 0.02));
        assert!(approx_eq(price, 0.988, 0.002));
    }

    #[test]
    fn test_constant_product_large_swap() {
        let (out, price) = constant_product_price(1000.0, 1000.0, 100.0, 0.003, true);
        // Larger trade = more slippage
        assert!(out < 100.0);
        assert!(price < 1.0);
    }

    #[test]
    fn test_constant_product_symmetric() {
        let (out1, _) = constant_product_price(1000.0, 1000.0, 50.0, 0.003, true);
        let (out2, _) = constant_product_price(1000.0, 1000.0, 50.0, 0.003, false);
        assert!(approx_eq(out1, out2, 0.01));
    }

    #[test]
    fn test_concentrated_liquidity() {
        let sqrt_price = 10.0;
        let sqrt_lower = 90.0_f64.sqrt();
        let sqrt_upper = 110.0_f64.sqrt();
        let liquidity = 1000.0;

        let (out, new_sqrt_p, impact) =
            concentrated_liquidity_price(liquidity, sqrt_price, sqrt_lower, sqrt_upper, 10.0, 0.003, true);

        assert!(out > 0.0);
        assert!(new_sqrt_p > sqrt_price); // Price increased
        assert!(impact >= 0.0);
    }

    #[test]
    fn test_price_conversions() {
        let sqrt_p = price_to_sqrt_price(100.0);
        assert!(approx_eq(sqrt_p, 10.0, 0.0001));

        let p = sqrt_price_to_price(10.0);
        assert!(approx_eq(p, 100.0, 0.0001));

        // Round trip
        let original = 12345.67;
        let sqrt_p = price_to_sqrt_price(original);
        let back = sqrt_price_to_price(sqrt_p);
        assert!(approx_eq(back, original, 0.01));
    }
}
